Update Unity AdMob Adapter for 4.0.0 release

vita-unity opened the pull request:
### Description
This PR updates our adapter to use our new 4.0.0 SDK and includes refactoring and improvements.

### How

Update Unity dependencies to 4.0.0.
Removed enablePerPlacementLoad flag from initialize call and any other references to classes removed in 4.0.0.
Refactor GADMAdNetworkAdapter and GADMediationAdapter to use proxy classes for UnityAds delegate.
Fix crash related to placementInUse map.
Small code improvements like comment updates and function restructuring.

### Test
I tested the adapter still works using our test app to load/show banners, interstitial and rewarded ads and ensuring all the events are properly sent

NOTE - Unity Ads 4.0 is not out yet, but wanted to start the review process so that the adapter can be updated shortly after our release

vita-unity commented:
I've signed the CLA, could you please re-run the check?

chlandacastro commented:
Hi, vita-unity,   the CLA still need to be signed.  
Regards, 
Christian.

chlandacastro commented on adapters/Unity/UnityAdapter/GADMAdapterUnityConstants.h:
```diff
@@ -14,6 +14,32 @@
 
 #import <Foundation/Foundation.h>
 
+typedef NS_ENUM(NSInteger, GADMAdapterUnityErrorCode) {
+  /// Invalid server parameters.
+  GADMAdapterUnityErrorInvalidServerParameters = 101,
+  /// Device not supported by UnityAds.
+  GADMAdapterUnityErrorDeviceNotSupported = 102,
+  /// UnityAds finished presenting with error state kUnityAdsFinishStateError.
+  GADMAdapterUnityErrorFinish = 103,
+  /// The Unity ad object is nil after calling its initializer.
+  GADMAdapterUnityErrorAdObjectNil = 104,
+  /// Failed to show Unity Ads due to ad not ready.
+  GADMAdapterUnityErrorShowAdNotReady = 105,
+  /// UnityAds called a placement changed callback with placement state
+  /// kUnityAdsPlacementStateNoFill.
+  GADMAdapterUnityErrorPlacementStateNoFill = 106,
+  /// UnityAds called a placement changed callback with placement state
+  /// kUnityAdsPlacementStateDisabled.
+  GADMAdapterUnityErrorPlacementStateDisabled = 107,
+  /// An ad was already loaded for this placement. UnityAds SDK does not support loading multiple
+  /// ads for the same placement.
+  GADMAdapterUnityErrorAdAlreadyLoaded = 108,
+  /// Banner size mismatch.
+  GADMAdapterUnityErrorSizeMismatch = 109,
+  /// UnityAds returned an initialization error
+  GADMAdapterUnityErrorAdInitializationFailure = 110
+};
+
```
The error codes should remain in the GADMediationAdapterUnity.h file   Those are not constants plus  we have to keep  consistency throughout all the adapters.

chlandacastro commented:
HI @vita-unity  the CLA is still missed.

vita-unity commented:
> HI @vita-unity the CLA is still missed.

Hi @chlandacastro, I'm checking the [cla page](https://cla.developers.google.com/clas) and I have the signed agreement 
![Screen Shot 2021-12-01 at 8 32 30 AM](https://user-images.githubusercontent.com/87673271/144274296-723322cf-27db-442e-bd0e-3e01d018b08e.png)
Do I need to sign something else?

chlandacastro commented:
hi @vita-unity.  It is still showing that the CLA is missing.   You need to be part of the Google Group tied to Unity's Corporate CLA  google-contributors@unity3d.com.  Could you check with Stephanie, from Unity,  about this missing CLA.
Regards, 
Christian.

joshh-devrel commented on adapters/Unity/UnityAdapter/AdNetworkAdapterProxy/GADMUnityBannerNetworkAdapterProxy.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 Google LLC.
```
For all the files that were created before this PR, the year should remain the same.

joshh-devrel replied:
For all the files that were created before this PR, the year should remain the same.
This applies to all the other files.

joshh-devrel commented on adapters/Unity/UnityAdapter/Categories/GADMediationConfiguration+Settings.h:
```diff
@@ -0,0 +1,29 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Nit: `+` sign in the file name is less preferable.

joshh-devrel commented on adapters/Unity/UnityAdapter/Categories/GADMediationConfiguration+Settings.m:
```diff
@@ -0,0 +1,41 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "GADMediationConfiguration+Settings.h"
```
Nit: `+` sign in the file name is less preferable.

vita-unity commented on adapters/Unity/UnityAdapter/AdNetworkAdapterProxy/GADMUnityBannerNetworkAdapterProxy.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 Google LLC.
+// Copyright 2021 Google LLC.
```
@chlandacastro github shows that this file (and a few more .h files) is renamed when actually it's a new file, so not sure if I should have 2020 here. Please let me know if I need to change this.

vita-unity commented:
@chlandacastro Thank you for your review! I've changed year back to 2020 for the old files and renamed categories files. Please let me know if you have any other comments.

chlandacastro commented:
@vita-unity  Hi  Vita!   can you  to sync this  PR with the latest changes.   This PR is already merged but it is not closed yet.  We want to confirm  the PR with the latest changes.
Thanks.

vita-unity commented:
@chlandacastro looks like latest commit didn't come through
